### PR TITLE
Program marketing slug to be URL of published Drupal page

### PR DESCRIPTION
**Pull request by ahsan-ul-haq**
ECOM-7310

### CODE REVIEW

File: course_discovery/apps/course_metadata/publishers.py [hunk 1]
```diff
@@ -292,65 +298,67 @@ def update_node_alias(self, obj, node_id, previous_obj):
         """
         new_alias = self.alias(obj)
         previous_alias = self.alias(previous_obj) if previous_obj else None
+        alias_list_url = self.alias_list_url(obj.marketing_slug)
```
Comment (schenedx) [inline course_discovery/apps/course_metadata/publishers.py]: I believe this line (301) and the next line (302) can be wrapped into a single function. It was repeated twice.

File: course_discovery/apps/course_metadata/publishers.py [hunk 2]
```diff
@@ -255,6 +256,11 @@ def publish_obj(self, obj, previous_obj=None):
                     self.edit_node(node_id, node_data)
 
             if node_id:
+                alias_list_url = self.alias_list_url(slugify(obj.title, allow_unicode=True))
```
Comment (schenedx) [inline course_discovery/apps/course_metadata/publishers.py]: This whole block (line 259 - line 262) can be wrapped into another new function. This is repeated twice. Once here, once on line 307 to line 311.

File: course_discovery/apps/course_metadata/publishers.py [hunk 3]
```diff
@@ -379,3 +387,9 @@ def alias_delete_path(self, url):
         delete_element = html.select('.delete.last a')
 
         return delete_element[0].get('href') if delete_element else None
+
+    def alias_list_url(self, slug):
```
Comment (schenedx) [inline course_discovery/apps/course_metadata/publishers.py]: #nit `def get_alias_list_url(self, slug)`

**Review by schenedx — CHANGES_REQUESTED** [review]
Mostly suggestions to keep the code dry and re-usable

**Comment by ahsan-ul-haq** [comment]
@rlucioni @MatthewPiatetsky could either of you please review this.

**Review by schenedx — APPROVED** [review]
👍  Looks good

File: course_discovery/apps/course_metadata/publishers.py [hunk 4]
```diff
@@ -292,65 +293,77 @@ def update_node_alias(self, obj, node_id, previous_obj):
...
 
-                    if response.status_code != 200:
-                        raise AliasDeleteError
+        if response.status_code != 200:
+            raise AliasDeleteError
 
     def alias(self, obj):
         return '{type}/{slug}'.format(type=obj.type.slug, slug=obj.marketing_slug)
 
-    @cached_property
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/publishers.py:345]: Why did you get rid of this?

File: course_discovery/apps/course_metadata/tests/test_publishers.py [hunk 5]
```diff
@@ -340,10 +345,13 @@ def test_update_node_alias(self):
         and deletes an old alias, if one existed, and that appropriate exceptions
         are raised for non-200 status codes.
         """
-        # No previous object is provided. Alias creation should occur, but alias
-        # deletion should not.
+        # No previous object is provided. Title Alias deletion and marketing
+        # slug Alias creation should occur, but slug alias deletion should not.
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/tests/test_publishers.py]: This tests the case where no previous object is provided. What is there to delete?

File: course_discovery/apps/course_metadata/publishers.py [hunk 6]
```diff
@@ -292,65 +293,77 @@ def update_node_alias(self, obj, node_id, previous_obj):
...
-                        'confirm': 1,
-                        'form_id': 'path_admin_delete_confirm',
-                        'op': 'Confirm',
-                    }
+        data = {
+            'confirm': 1,
+            'form_id': 'path_admin_delete_confirm',
+            'op': 'Confirm'
+        }
+        data.update(self.alias_form_inputs(alias_delete_url))
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/publishers.py]: Why did you change the `**` syntax that was being used before?

File: course_discovery/apps/course_metadata/publishers.py [hunk 7]
```diff
@@ -292,65 +293,77 @@ def update_node_alias(self, obj, node_id, previous_obj):
...
             }
 
             data = {
-                **self.alias_form_inputs,
                 'alias': new_alias,
                 'form_id': 'path_admin_form',
                 'op': 'Save',
                 'source': 'node/{}'.format(node_id),
             }
+            data.update(self.alias_form_inputs(self.alias_add_url))
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/publishers.py]: Why did you change the `**` syntax that was being used?

File: course_discovery/apps/course_metadata/publishers.py [hunk 8]
```diff
@@ -292,65 +293,77 @@ def update_node_alias(self, obj, node_id, previous_obj):
         """
         new_alias = self.alias(obj)
         previous_alias = self.alias(previous_obj) if previous_obj else None
+        new_alias_delete_path = self.alias_delete_path(self.get_alias_list_url(obj.marketing_slug))
 
-        if new_alias != previous_alias:
-            alias_add_url = '{}/add'.format(self.alias_api_base)
+        if new_alias != previous_alias or not new_alias_delete_path:
+            # Delete old alias before saving the new one.
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/publishers.py:299]: Is it going to be a problem if the old alias is deleted before the new one is created?

File: course_discovery/apps/course_metadata/publishers.py [hunk 9]
```diff
@@ -255,6 +255,7 @@ def publish_obj(self, obj, previous_obj=None):
                     self.edit_node(node_id, node_data)
 
             if node_id:
+                self.get_and_delete_alias(slugify(obj.title, allow_unicode=True))
```
Comment (rlucioni) [inline course_discovery/apps/course_metadata/publishers.py]: Is the `allow_unicode` kwarg necessary here?

**Review by rlucioni — CHANGES_REQUESTED** [review]

File: course_discovery/apps/course_metadata/publishers.py (hunk 4 above)
Comment (ahsan-ul-haq) [reply course_discovery/apps/course_metadata/publishers.py:345]: I don't think we can cache this property as form inputs are being scraped based on the url provided i.e. alias_delete_url or alias_add_url

File: course_discovery/apps/course_metadata/tests/test_publishers.py (hunk 5 above)
Comment (ahsan-ul-haq) [reply course_discovery/apps/course_metadata/tests/test_publishers.py]: For a newly created node in drupal, when it is published an alias based on the title of the node is being created we need to delete that alias and then create a marketing slug based alias. Here title based alias is being deleted.

File: course_discovery/apps/course_metadata/publishers.py (hunk 8 above)
Comment (ahsan-ul-haq) [reply course_discovery/apps/course_metadata/publishers.py:299]: I observe an issue if we create a new alias and then delete the old one it does not create the new alias just create an URL redirect which is being resolved by first deleting the old one and then create a new one.

**Comment by ahsan-ul-haq** [comment]
@rlucioni could you please review this.

File: course_discovery/apps/course_metadata/tests/test_publishers.py (hunk 5 above)
Comment (rlucioni) [reply course_discovery/apps/course_metadata/tests/test_publishers.py]: @ahsan-ul-haq please make that more clear in this comment for others who may not know this.

**Review by rlucioni — APPROVED** [review]

**Comment by rlucioni** [comment]
@ahsan-ul-haq please remember to squash your commits!

**Pull request merged**